internal/agent/agent_syncer.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2023 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package agent
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	apiv1 "github.com/numaproj-labs/numaplane/api/v1alpha1"
+	kvsource "github.com/numaproj-labs/numaplane/internal/keyvaluegenerator"
+	//kubeutil "github.com/argoproj/gitops-engine/pkg/utils/kube"
+	//"k8s.io/client-go/rest"
+	//"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type AgentSyncer struct {
+	logger *zap.SugaredLogger
+
+	// the source where we watch manifests
+	gitSource *apiv1.CredentialedGitSource
+
+	// this is the source of our key/value pairs
+	kvSource kvsource.KVSource
+
+	// current Config file
+	config AgentConfig
+
+	// keep track of revision number of config file so we know when it's new
+	configRevision int
+	// todo: add all of these in
+	/*
+		client    client.Client
+		config    *rest.Config
+		rawConfig *rest.Config
+		kubectl   kubeutil.Kubectl
+
+		stateCache  LiveStateCache
+	*/
+}
+
+func NewAgentSyncer(logger *zap.SugaredLogger) *AgentSyncer {
+	return &AgentSyncer{
+		logger:         logger,
+		configRevision: -1, // setting this < 0 enables us to check it initially
+	}
+}
+
+// Run function runs in a loop, syncing the manifest if it's changed, checking every x seconds
+// It responds to all of the following events:
+// 1. Config change;
+// 2. Source Manifest change;
+// 3. If file generator is used, then a change from that file
+func (syncer *AgentSyncer) Run(ctx context.Context) {
+
+	for {
+		select {
+		default:
+
+			// Determine the latest value of our GitSource definition
+			syncer.evaluateGitSource()
+
+			// fetch the GitSource and apply the resources
+			syncer.syncLatest()
+
+			time.Sleep(time.Duration(syncer.config.TimeIntervalSec) * time.Second)
+		case <-ctx.Done():
+			syncer.logger.Info("context ended, terminating AgentSyncer watch")
+			return
+		}
+	}
+
+}
+
+// determine if Config was updated and if so, get latest
+// return if new
+func (syncer *AgentSyncer) checkConfigUpdate() bool {
+	configManager := GetConfigManagerInstance()
+	var err error
+	var newRevision int
+	// Reload our copy of the Config if it changed (or load it the first time upon starting)
+	if configManager.GetRevisionIndex() > syncer.configRevision {
+		syncer.config, newRevision, err = configManager.GetConfig()
```

Comment:
Why can't we just call GetConfig each time when we need the config? it will be always the latest in that case.

Reply:
For one, it would mean creating a deep copy of the Config, which we don't need to do. But moreover, it would mean having to create a new KVSource based on the generator, which could potentially be time-intensive, such as the case of the `MultiClusterFileKVSource` might be (or if it's not, then another future one could be).